src/python/pants/backend/native/subsystems/binaries/gcc.py

```diff
@@ -7,64 +7,81 @@
 import os
 
 from pants.backend.native.config.environment import (CCompiler, CppCompiler, GCCCCompiler,
-                                                     GCCCppCompiler)
-from pants.backend.native.subsystems.utils.parse_search_dirs import ParseSearchDirs
+                                                     GCCCppCompiler, Platform)
 from pants.binaries.binary_tool import NativeTool
 from pants.engine.rules import RootRule, rule
 from pants.engine.selectors import Select
-from pants.util.memo import memoized_method, memoized_property
-from pants.util.strutil import create_path_env_var
+from pants.util.memo import memoized_method
 
 
 class GCC(NativeTool):
   options_scope = 'gcc'
   default_version = '7.3.0'
   archive_type = 'tgz'
 
-  @classmethod
-  def subsystem_dependencies(cls):
-    return super(GCC, cls).subsystem_dependencies() + (ParseSearchDirs.scoped(cls),)
-
-  @memoized_property
-  def _parse_search_dirs(self):
-    return ParseSearchDirs.scoped_instance(self)
-
-  def _lib_search_dirs(self, compiler_exe, path_entries):
-    return self._parse_search_dirs.get_compiler_library_dirs(
-      compiler_exe,
-      env={'PATH': create_path_env_var(path_entries)})
-
   @memoized_method
   def path_entries(self):
     return [os.path.join(self.select(), 'bin')]
 
-  def c_compiler(self):
-    exe_filename = 'gcc'
-    path_entries = self.path_entries()
+  _PLATFORM_INTERMEDIATE_DIRNAME = {
```

Comment:
So this is actually only a path into the specific gcc archive we provide, which is why I was thinking it was ok to lean on. However, the `ParseSearchDirs` approach actually may be able to get this -- doing it manually results in:
```bash
> PATH="${HOME}/.cache/pants/bin/gcc/mac/10.13/7.3.0/gcc/bin" g++ -print-search-dirs | grep -P '^libraries' | sed -re 's#^libraries: =##g' | tr ':' '\n' | sort | uniq | parallel -n1 readlink -f
/Users/dmcclanahan/.cache/pants/bin/gcc/mac/10.13/7.3.0/gcc/lib/gcc/x86_64-apple-darwin17.5.0/7.3.0
/Users/dmcclanahan/.cache/pants/bin/gcc/mac/10.13/7.3.0/gcc/lib
/Users/dmcclanahan/.cache/pants/bin/gcc/mac/10.13/7.3.0/gcc/lib/gcc
```

This looks like it can be manipulated to find the include dirs we want instead of generating those paths ourselves, which would allow us to revert many of the changes to this file.

Reply:
OK, that would be excellent.

But this clarification also points to this issue https://github.com/pantsbuild/binaries/issues/78; ie: x86_64-apple-darwin17.5.0 is a bit specific, which would be OK if it corresponded to 10.11.

Reply:
The fact that this entire class refers to the specifics of our GCC build is worth pointing out in a class level doc most likely.

Reply:
Yes, 100% agreed.

Reply:
To clarify, `17.5.0` is from the LLVM binary release for OSX -- this is the package we provide for all OSX (because it's the package LLVM provides for all OSX), and is not an artifact of any build process we perform ourselves. I am making the `ParseSearchDirs` approach work regardless, just wanted to clear up why I thought that path was ok.

Reply:
I was having difficulty getting the `ParseSearchDirs` solution to work, so e67044c scraps that entirely and uses a helper method `_get_check_single_path_by_glob()` which is used to locate directories within our provided gcc distribution.

I was annoyed that we couldn't just invoke the binary -- although there may be a better way in the future, adding `/lib` and `/include` to directory paths (as I mentioned could be doable above) ended up getting way too long and needlessly complex.

Reply:
The process of using globs was split off into another `pants.backend.native.utils` subsystem named `ArchiveFileMapper`, and docstrings describing that `GCC` and `LLVM` rely on the current specific layout of the archive were added as well, in 3051235f8efbc80f183b23f30ae414851d97af37.